simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp

```diff
@@ -85,6 +85,50 @@ HdMapUtils::HdMapUtils(
     lanelet::utils::query::shoulderLanelets(lanelet::utils::query::laneletLayer(lanelet_map_ptr_));
 }
 
+auto HdMapUtils::gelAllCanonicalizedLaneletPoses(
+  const traffic_simulator_msgs::msg::LaneletPose & lanelet_pose) const
+  -> std::vector<traffic_simulator_msgs::msg::LaneletPose>
+{
+  std::vector<traffic_simulator_msgs::msg::LaneletPose> canonicalized_all;
+  if (lanelet_pose.s < 0) {
+    if (const auto ids = getPreviousLaneletIds(lanelet_pose.lanelet_id); ids.empty()) {
+      return canonicalized_all;
+    } else {
+      for (const auto id : ids) {
+        const auto lanelet_pose_tmp = traffic_simulator::helper::constructLaneletPose(
+          id, lanelet_pose.s + getLaneletLength(id), lanelet_pose.offset);
+        if (auto canonicalized_lanelet_poses = gelAllCanonicalizedLaneletPoses(lanelet_pose_tmp);
+            canonicalized_lanelet_poses.empty()) {
+          canonicalized_all.emplace_back(lanelet_pose_tmp);
+        } else {
+          std::copy(
+            canonicalized_lanelet_poses.begin(), canonicalized_lanelet_poses.end(),
+            std::back_inserter(canonicalized_all));
+        }
+      }
+    }
+  }
+  if (lanelet_pose.s > (getLaneletLength(lanelet_pose.lanelet_id))) {
+    if (const auto ids = getNextLaneletIds(lanelet_pose.lanelet_id); ids.empty()) {
+      return canonicalized_all;
+    } else {
+      for (const auto id : ids) {
+        const auto lanelet_pose_tmp = traffic_simulator::helper::constructLaneletPose(
+          id, lanelet_pose.s - getLaneletLength(lanelet_pose.lanelet_id), lanelet_pose.offset);
+        if (auto canonicalized_lanelet_poses = gelAllCanonicalizedLaneletPoses(lanelet_pose_tmp);
+            canonicalized_lanelet_poses.empty()) {
+          canonicalized_all.emplace_back(lanelet_pose_tmp);
+        } else {
+          std::copy(
+            canonicalized_lanelet_poses.begin(), canonicalized_lanelet_poses.end(),
+            std::back_inserter(canonicalized_all));
+        }
+      }
+    }
+  }
+  return canonicalized_all;
```

Comment:
We believe that this writing style makes the structure of the function very difficult to grasp, since the If in lines 93 and 111 appear to be something that cannot be established at the same time.
We suggest that you rewrite the function as you see fit.

Reply:
```suggestion
  /// @note Define lambda functions for canonicalizing previous/next lanelet.
  const auto canonicalize_to_previous_lanelet =
    [this](const auto & lanelet_pose) -> std::vector<traffic_simulator_msgs::msg::LaneletPose> {
    if (const auto ids = getPreviousLaneletIds(lanelet_pose.lanelet_id); !ids.empty()) {
      std::vector<traffic_simulator_msgs::msg::LaneletPose> canonicalized_all;
      for (const auto id : ids) {
        const auto lanelet_pose_tmp = traffic_simulator::helper::constructLaneletPose(
          id, lanelet_pose.s + getLaneletLength(id), lanelet_pose.offset);
        if (const auto canonicalized_lanelet_poses =
              gelAllCanonicalizedLaneletPoses(lanelet_pose_tmp);
            canonicalized_lanelet_poses.empty()) {
          canonicalized_all.emplace_back(lanelet_pose_tmp);
        } else {
          std::copy(
            canonicalized_lanelet_poses.begin(), canonicalized_lanelet_poses.end(),
            std::back_inserter(canonicalized_all));
        }
      }
      return canonicalized_all;
    } else {
      return {};
    }
  };
  const auto canonicalize_to_next_lanelet =
    [this](const auto & lanelet_pose) -> std::vector<traffic_simulator_msgs::msg::LaneletPose> {
    if (const auto ids = getNextLaneletIds(lanelet_pose.lanelet_id); !ids.empty()) {
      std::vector<traffic_simulator_msgs::msg::LaneletPose> canonicalized_all;
      for (const auto id : ids) {
        const auto lanelet_pose_tmp = traffic_simulator::helper::constructLaneletPose(
          id, lanelet_pose.s - getLaneletLength(lanelet_pose.lanelet_id), lanelet_pose.offset);
        if (const auto canonicalized_lanelet_poses =
              gelAllCanonicalizedLaneletPoses(lanelet_pose_tmp);
            canonicalized_lanelet_poses.empty()) {
          canonicalized_all.emplace_back(lanelet_pose_tmp);
        } else {
          std::copy(
            canonicalized_lanelet_poses.begin(), canonicalized_lanelet_poses.end(),
            std::back_inserter(canonicalized_all));
        }
      }
      return canonicalized_all;
    } else {
      return {};
    }
  };

  /// @note If s value unders 0, it means this pose is on the previous lanelet.
  if (lanelet_pose.s < 0) {
    return canonicalize_to_previous_lanelet(lanelet_pose);
  }
  /// @note If s value overs it's lanelet length, it means this pose is on the next lanelet.
  else if (lanelet_pose.s > (getLaneletLength(lanelet_pose.lanelet_id))) {
    return canonicalize_to_next_lanelet(lanelet_pose);
  }
  /// @note If s value is in range [0,length_of_the_lanelet], return empty.
  else {
    /// Review comment, please remove before next code review. => Is this return value is valid? I think {lanelet_pose}; is a valid return value.
    return {};
  }
```

Reply:
I left review comment `Review comment, please remove before next code review. => Is this return value is valid? I think {lanelet_pose}; is a valid return value.`. Please check.

Reply:
I've applied the suggested change -> d88d8b60c53d2c3e5cf66cb691f802ad4812b507
I've also added additional testcases for s in range [s, length_of_lanelet]